00XX-efficient-msms-for-non-native-pickles-verification.md

```diff
@@ -0,0 +1,481 @@
+- [Projective](https://www.hyperelliptic.org/EFD/g1p/auto-shortw-projective.html)
+
+Again, **in practice** affine coordinates will be most likely the most efficient, and also easy to implement, so the suggested path is to start by implementing those, and revisiting the technique much later in the project timeline.
+
+
+We might find inspiration and decide which encoding is more efficient by looking at existing FF/ECC implementations:
+- https://github.com/privacy-scaling-explorations/halo2wrong
+- https://github.com/DelphinusLab/halo2ecc-s
+- https://hackmd.io/@arielg/B13JoihA8
+
+
+### Algorithm Performance, Circuit Layout, and Multi-Circuit Folding
+
+The circuit that we are implementing is large, and its layout should be carefully considered. Luckily, we have some liberty in design decisions -- the algorithm needs to be implemented on a *variant* of Kimchi which can allow long rows, additive lookups (memory access), and support for folding. Two last features are (being) developed in the zkVM project.
+
+Note that each iteration (one run) of the sub-MSM algorithm with limited buckets takes $3n$ additions at most:
+- The cycle in the beginning populates the buckets and it takes $n$ additions because that's how much multiplications are there in `to_scale_pairs`.
+- The end loop does two additions per each iteration (and each iteration assumes a non-zero bucket, assume there are $n'$ of these buckets), so we have $2 \cdot n'$ additions. Since $n' < n$, the worst case is $2 n$ additions.
+
+In addition to $3n$ addition per sub-MSM run, we will need to sum all the results of it. Luckily, this computation is relatively cheap --- we require one extra addition per run.
+
+In total, the whole algorithm then requires $3n \cdot l$ additions because we need to run the inner algorithm $l$ times, and then sum the results (which is negligibly small and can be accumulated along the way). Assuming worst (of the two) case of $n = 2^{16}$ and $k = 15$, we get about $2^{17}$ additions, which is still more than $2^{16}$ budget. But even with the easier $n = 2^{15}$ we still have to fit $3$ times more FF EC additions than there are rows available.
+
+
+With that in mind, we will use folding to split the total computation into chunks that fit into our $n$-element SRS. Each folding repetition will externally look like an elliptic curve addition w.r.t. the total accumulator. We will elaborate on the design of a concrete circuit later, but so far we can assume that the only shared states between the rounds of folding are (1) total accumulator for the computed value $\sum\limits_{j=1}^i \mathsf{subres}_j$, (2) RAM lookups. Note that the RAM lookups are not ZK and they don't need to be.
+
+
+Now, there are several approach that can be taken in terms of a concrete circuit design, and deciding on the approach is part of the implementation effort itself since it is (arguably) too complicated for the RFC. The approaches are as follows:
```

Comment:
> Now, there are several approach that can be taken in terms of a concrete circuit design, and deciding on the approach is part of the implementation effort itself since it is (arguably) too complicated for the RFC.

This is the *whole point* of the RFC. Delete.